src/asset-management/EulerV2Manager.sol

```diff
@@ -305,17 +274,50 @@ contract AaveManager is IAssetManager, Owned(msg.sender), ReentrancyGuard {
 
-    function claimRewardForMarket(address aMarket, address aReward)
+    function claimRewards(
+        IDistributor aDistributor,
+        address[] calldata aUsers,
+        address[] calldata aTokens,
+        uint256[] calldata aAmounts,
+        bytes32[][] calldata aProofs
+    ) external onlyGuardianOrOwner {
+        aDistributor.claim(aUsers, aTokens, aAmounts, aProofs);
+
+        for (uint256 i = 0; i < aTokens.length; ++i) {
+            SafeTransferLib.safeTransfer(aTokens[i], msg.sender, aAmounts[i]);
+        }
+    }
+
+    /// @dev The guardian or owner would first claim rewards on behalf of the asset manager by calling `claimRewards` and sell it for the underlying token.
+    /// The asset manager pulls the assets, deposits it to the vault, and distribute the proceeds in the form of ERC4626 shares to the pairs.
+    /// Due to integer arithmetic the last pair of the array will get one or two more shares, so as to maintain the invariant that
+    /// the sum of shares for all pair+token equals the totalShares.
+    function distributeRewardForPairs(IERC20 aAsset, uint256 aAmount, IAssetManagedPair[] calldata aPairs)
         external
         onlyGuardianOrOwner
-        returns (uint256 rClaimed)
     {
-        require(aReward != address(0), "AM: REWARD_TOKEN_ZERO");
-        require(aMarket != address(0), "AM: MARKET_ZERO");
-
-        address[] memory lMarkets = new address[](1);
-        lMarkets[0] = aMarket;
+        // pull assets from guardian / owner
+        SafeTransferLib.safeTransferFrom(address(aAsset), msg.sender, address(this), aAmount);
+        IERC4626 lVault = assetVault[aAsset];
+        SafeTransferLib.safeApprove(address(aAsset), address(lVault), aAmount);
+        uint256 lNewShares = lVault.deposit(aAmount, address(this));
+
+        uint256 lOldTotalShares = totalShares[lVault];
+        totalShares[lVault] = lOldTotalShares + lNewShares;
+        uint256 lSharesAllocated;
+        uint256 lLength = aPairs.length;
+        for (uint256 i = 0; i < lLength - 1; ++i) {
+            uint256 lOldShares = shares[aPairs[i]][aAsset];
+            uint256 lNewSharesEntitled = lNewShares.mulDiv(lOldShares, lOldTotalShares);
+            shares[aPairs[i]][aAsset] = lOldShares + lNewSharesEntitled;
+            lSharesAllocated += lNewSharesEntitled;
+        }
 
-        rClaimed = rewardsController.claimRewards(lMarkets, type(uint256).max, msg.sender, aReward);
+        // the last in the list will take all the remaining shares, and sometimes will get 1 or 2 more than they're entitled to
+        // due to the rounding down in previous calculations for other pairs
+        // this is to prevent the sum of each pair+token's shares not summing up to totalShares
+        shares[aPairs[lLength - 1]][aAsset] = lNewShares - lSharesAllocated;
+        lSharesAllocated += shares[aPairs[lLength - 1]][aAsset];
+        require(lSharesAllocated == lNewShares, "AM: REWARD_SHARES_MISMATCH");
```

Comment:
Why not just have the interface be a single pair, what value do we get from this multi pair support (cause we take a lot of risk, for instance this could fail if we get unlucky with rounding).